src/lib/gadgets/elliptic-curve.ts

```diff
@@ -0,0 +1,575 @@
+ *
+ * Note: this algorithm targets a small number of points, like 2 needed for ECDSA verification.
+ *
+ * TODO: could use lookups for picking precomputed multiples, instead of O(2^c) provable switch
+ * TODO: custom bit representation for the scalar that avoids 0, to get rid of the degenerate addition case
+ * TODO: glv trick which cuts down ec doubles by half by splitting s*P = s0*P + s1*endo(P) with s0, s1 in [0, 2^128)
+ */
+function multiScalarMul(
+  Curve: CurveAffine,
+  scalars: Field3[],
+  points: Point[],
+  tableConfigs: (
+    | { windowSize?: number; multiples?: Point[] }
+    | undefined
+  )[] = [],
+  ia?: point
+): Point {
+  let n = points.length;
+  assert(scalars.length === n, 'Points and scalars lengths must match');
+  assertPositiveInteger(n, 'Expected at least 1 point and scalar');
+
+  // constant case
+  if (scalars.every(Field3.isConstant) && points.every(Point.isConstant)) {
+    // TODO dedicated MSM
+    let s = scalars.map(Field3.toBigint);
+    let P = points.map(Point.toBigint);
+    let sum = Curve.zero;
+    for (let i = 0; i < n; i++) {
+      sum = Curve.add(sum, Curve.scale(P[i], s[i]));
+    }
+    return Point.from(sum);
+  }
+
+  // parse or build point tables
+  let windowSizes = points.map((_, i) => tableConfigs[i]?.windowSize ?? 1);
+  let tables = points.map((P, i) =>
+    getPointTable(Curve, P, windowSizes[i], tableConfigs[i]?.multiples)
+  );
+
+  // slice scalars
+  let b = Curve.order.toString(2).length;
+  let scalarChunks = scalars.map((s, i) =>
+    slice(s, { maxBits: b, chunkSize: windowSizes[i] })
+  );
+
+  ia ??= initialAggregator(Curve);
+  let sum = Point.from(ia);
+
+  for (let i = b - 1; i >= 0; i--) {
+    // add in multiple of each point
+    for (let j = 0; j < n; j++) {
+      let windowSize = windowSizes[j];
+      if (i % windowSize === 0) {
+        // pick point to add based on the scalar chunk
+        let sj = scalarChunks[j][i / windowSize];
+        let sjP =
+          windowSize === 1
+            ? points[j]
+            : arrayGetGeneric(Point.provable, tables[j], sj);
+
+        // ec addition
+        let added = add(sum, sjP, Curve.modulus);
+
+        // handle degenerate case (if sj = 0, Gj is all zeros and the add result is garbage)
+        sum = Provable.if(sj.equals(0), Point.provable, sum, added);
+      }
+    }
+
+    if (i === 0) break;
+
+    // jointly double all points
+    // (note: the highest couple of bits will not create any constraints because sum is constant; no need to handle that explicitly)
+    sum = double(sum, Curve.modulus);
+  }
+
+  // the sum is now 2^(b-1)*IA + sum_i s_i*P_i
+  // we assert that sum != 2^(b-1)*IA, and add -2^(b-1)*IA to get our result
+  let iaFinal = Curve.scale(Curve.fromNonzero(ia), 1n << BigInt(b - 1));
+  Provable.equal(Point.provable, sum, Point.from(iaFinal)).assertFalse();
+  sum = add(sum, Point.from(Curve.negate(iaFinal)), Curve.modulus);
+
+  return sum;
```

Comment:
you mean on `sum`? it's has to be in the subgoup and on the curve because it was created by scaling `P` and `G`

Reply:
i meant on the pubkey.

Reply:
this is the multiscalarmul method, it doesn't know about a pubkey. In the ecdsa gadget I don't do it because it will be part of checks on inputs when they are created, done automatically in a higher-level interface I'm adding on top of this: https://github.com/o1-labs/o1js-bindings/pull/215